internal/ast/compiler/disjunctions_infer_mapping.go

```
@@ -0,0 +1,261 @@
+	}
+
+	if def.Kind == ast.KindMap {
+		return pass.processMap(schema, def)
+	}
+
+	if def.Kind == ast.KindStruct {
+		return pass.processStruct(schema, def)
+	}
+
+	if def.Kind == ast.KindDisjunction {
+		return pass.processDisjunction(schema, def)
+	}
+
+	return def, nil
+}
+
+func (pass *DisjunctionInferMapping) processArray(schema *ast.Schema, def ast.Type) (ast.Type, error) {
+	var err error
+
+	def.Array.ValueType, err = pass.processType(schema, def.AsArray().ValueType)
+	if err != nil {
+		return ast.Type{}, err
+	}
+
+	return def, nil
+}
+
+func (pass *DisjunctionInferMapping) processMap(schema *ast.Schema, def ast.Type) (ast.Type, error) {
+	var err error
+
+	def.Map.ValueType, err = pass.processType(schema, def.AsMap().ValueType)
+	if err != nil {
+		return ast.Type{}, err
+	}
+
+	return def, nil
+}
+
+func (pass *DisjunctionInferMapping) processStruct(schema *ast.Schema, def ast.Type) (ast.Type, error) {
+	var err error
+
+	for i, field := range def.Struct.Fields {
+		def.Struct.Fields[i].Type, err = pass.processType(schema, field.Type)
+		if err != nil {
+			return ast.Type{}, errors.Join(
+				fmt.Errorf("could not process struct field '%s'", field.Name),
+				err,
+			)
+		}
+	}
+
+	return def, nil
+}
+
+func (pass *DisjunctionInferMapping) processDisjunction(schema *ast.Schema, def ast.Type) (ast.Type, error) {
+	var err error
+
+	if !def.Disjunction.Branches.HasOnlyRefs() {
+		return def, nil
+	}
+
+	def.Disjunction, err = pass.ensureDiscriminator(schema, def.Disjunction)
+	if err != nil {
+		return ast.Type{}, err
+	}
+
+	return def, nil
+}
+
+func (pass *DisjunctionInferMapping) ensureDiscriminator(schema *ast.Schema, def *ast.DisjunctionType) (*ast.DisjunctionType, error) {
+	// discriminator-related data was set during parsing: nothing to do.
+	if def.Discriminator != "" && len(def.DiscriminatorMapping) != 0 {
+		return def, nil
+	}
+
+	if def.Discriminator == "" {
+		def.Discriminator = pass.inferDiscriminatorField(schema, def)
```

Shouldn't we return an error if we don't find the discriminator?

---

All the error handling is done in the `buildDiscriminatorMapping()` function called a few lines below